### ISSUE TITLE
`vue/v-if-else-key` false-positives and false-negatives

### ISSUE THREAD

**Issue by FloEdelmann** [vuejs/eslint-plugin-vue#2342]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 8.55.0
- **eslint-plugin-vue version:** 9.19.2
- **Node version:** 18.18.2
- **Operating System:** Ubuntu

**Please show your full configuration:**
```json5
{
    "rules": {
        "vue/v-if-else-key": "error"
    }
}
```

**What did you do?**
**What did you expect to happen?**
**What actually happened?**
**Repository to reproduce this issue**

→ please see the following demo links, where each issue is highlighted:

* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJylUUsOgjAQvco4GzY2JC4J5SRsKgxIhELaghpC4mk8mCexBDRVjMbYVd+b9n3aHpM6JQwwNFQ1pTAUxRLsCk1R3vcTnufQsSLjMWozAs45eFldezE6hx8CTFNekTTR9XwJ/ScGwhVjQMeGEkMpCDgIJQuZw44UAWNv5ayTIeWozYST038psshiC1CpadFiK5RtAXs6WV7YPl8zy/pD6N89t395TtcscF4A16jakjQGPXYt+ePXTUGsIwabYbgBPbajsw==
* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJxtj8EKgzAQRH9lu+cugR4l9Uu8WB3bUE0kRq2I/97USpHinnYHZt7szIUrwQnrgKat84A0sxRHl2bY1oNzlZpJCte0zsIGGsRU14wr5zImlZI+iRBeLYqAkqyjMffW2Ds94EEiKf2lRUAMQd0h2ndg9SXvlCPuLfcrdyu/mbT6fcVn9n2NjpOZhx7q41tp8sTEyWVZ3uUzTjQ=
* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJx1jksOgzAMRK/iel0rUpco5SRsKAwtKiQoBChC3L1p1E829Wpsye/NxpWtwRlrj37oSo+8MBRG1+38jnHtV6lsP1gD42mWtjkX3FhbMKmc9EGE8BhQedRkLC2lM6250g0OJJKCAjf8oxuRXP8qLqWLigSgfs0+Watvez6ymzqMnG08T1AvTrTJHStnp31/AoNnSPQ=
* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJyFkMsOgjAQRX9lnLUNiUuCfEk3BS5KhJaUhyLh3y0gkSDGrqaTmzMnt+fYJGCfgxpFmasaodTkXpBk7XucvkUnYlOURkPX1IosPUtOjZFMXkjBQQjCo0RcIyFt6K6szvSFrrAgIRxohXJkR0BeYcZEykpe3ZpCi85yK1LPr9BPsWhfTO14/eONmo60sfM2bc3LT2XLvArykW2To2K/57aBN6rOJdzQsX8ahhedG3Nr
* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJyFj0EOgjAQRa8yztqmiUtSPQmbSj/aWFpSCoqEu4uALHFWM8nLf/MHLoIBZ6wSqtrphEvuaRplbLeu81n1oghVHTx8ok7Y8pxzGULOJCfsD3jVcQZJHYQgvGoUCYY0PXX01t/ojggSYkcI1+AX9l6te+wGKLk0UXJryEeOrUPD2cBdC/n9chE80HN2GscPff5WVQ==
* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJzVk80OgjAMx1+l9uyyxCOZPMkuA4oSYVvGwA/Cu4uiHhQFTTSxp7Zp/vn1q8HYJIQBCk+FzZWnUGroTCRZfXEHwnMqNoU1mrRnCmqWpUuJqTESgYcgZowB7SzFnhLQBrbK6UyvYE2OgLHnWlGnRXlJvWCk3NuCgo+zP6bGWnpZfQ99+GwK/F9Jp4x8MvEA7Xcu4JcYV1/w26PhHF2VU4lBg3VF/ETT72ZDewwWbXsEm/H/lA==
* https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJylUcsOgjAQ/JV1z1YSj6T6JVwKLNoILSkPH4R/t+WRIIKJ2ku7uzOTmW6DkY4JfeQlZXkqSjoGCuzhsayH51hCzWRyCDDROsDJrJuPdAtKtLEoaTsgFbi7eMPPNR0qQPAvdB+qnYwXWXNmKIwlemvISGe5VqRKJiyD0oIGmgthacA3jAHdcopKikFpuAqjpDrBmQwBY5+EqRMOB+Ef1bj3+tV9b7aOJeB6NvH4Ott/eUZrE9+4RVOlVKDfYF2R59bVe7Q7Rn/ftk/X47Qg

CC @felipemelendez

**Comment by felipemelendez** [comment]
Hi, I'm sorry to see this, I hope we can resolve this together.

Here is the solution to most of the problems highlighted here:
At some point the check to ensure that all components were of the same type was removed, so perhaps we ought to reinstate that. Perhaps we could do something like this:
```
/**
 * Checks whether a conditional family has formed.
 * A conditional family is considered formed if all its components should be of the same type,
 * and it must have an 'if' condition and either an 'else' condition or at least one 'else-if' condition.
 *
 * @param {ConditionalFamily} family - The conditional family object with 'if', 'else-if', and 'else' properties.
 * @returns {boolean} True if a conditional family has formed, false otherwise.
 */
const hasFormedConditionalFamily = (family) => {
  const familyComponentName = family.if.rawName

  if (family.else && family.else.rawName !== familyComponentName) {
    return false
  }
  if (family.elseIf.some((node) => node.rawName !== familyComponentName)) {
    return false
  }

  return Boolean(family.else || family.elseIf.length > 0)
}
```

**Comment by felipemelendez** [comment]
Not sure about the rest, any ideas?

**Comment by felipemelendez** [comment]
hmm, for some reason that component name check change I just suggested solved all the problems that Flo highlighted that are related to mixed components getting targetted but causes a problem with this scenario:
```
<div>
  <ComponentA v-if="foo" />
  <ComponentB v-else-if="bar" />
  <ComponentA v-else />
</div>
```

That transition from A to B to A causes the rule to skp the A's altogether... I'll investigate that as soon as I get a chance
